_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,421 @@
+        ├╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:110)
+        ├╴call ERC20::balanceOf(address)(address) (src/Solidity.t.sol:104)
+        │  └╴← (uint256)
+        ├╴log_bytes32(bytes32) (lib/ds-test/src/test.sol:108)
+        ├╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:109)
+        └╴log_named_uint(bytes32, uint256) (lib/ds-test/src/test.sol:110)
+```
+
+Looking into the output, we can see that this represents the case where `dst` is the same as the
+sender (in this case the test contract).
+
+In this case the counter example found doesn't represent a bug in the implementation of `transfer`,
+but rather shows that our understanding of the expected behaviour was flawed: an exhaustive
+description of the behaviour of `transfer` must take self transfers into account. This kind of
+situation is common when applying formal methods, where we are forced to consider all possible edge
+cases.
+
+It is also worth noting that fuzzing would be very unlikely to catch this edge case: there are 2^20
+possible addresses, and the chance that a randomly generated address would match the address of the
+test contract is miniscule. You can try it out yourself by renaming the `prove_transfer` method to
+`test_transfer` and seeing if a counter example is found.
+
+A test for `transfer` that takes self-transfers into account could look like this:
+
+```solidity
+function prove_transfer(address dst, uint amt) public {
+    log_named_address("this", address(this));
+
+    uint preBalThis = token.balanceOf(address(this));
+    uint preBalDst  = token.balanceOf(dst);
+
+    token.transfer(dst, amt);
+
+    // no change for self-transfer
+    uint delta = dst == address(this) ? 0 : amt;
+
+    // balance of this has been reduced by `delta`
+    assertEq(sub(preBalThis, token.balanceOf(address(this))), delta);
+
+    // balance of dst has been increased by `delta`
+    assertEq(sub(token.balanceOf(dst), preBalDst), delta);
+}
+```
+
+## Execution Environment And Limits to Proof
+
+In order to understand the limits of the proofs that it is possible to produce using this framework,
+an understanding of the environment in which they are run is essential:
+
+- All variables in the environment (e.g. caller, gas, timestamp) remain concrete
+- All storage slots are set to zero at the beginning of the tests (meanings tests are effectively run against an empty blockchain)
+
+In fact, the only symbolic variables introduced into the test environment are those that are
+specified in the signature of the test method. This means that the proofs are exhaustive *only over
+the input variables*. As an example, consider the `prove_transfer` test from the example above. The
+`totalSupply` is always `uint(-1)`, and the test would not catch obviously faulty implementations of
+`transfer` like the one below:
+
+```solidity
+function transfer(address dst, uint amt) public {
+    require(totalSupply > uint56(-1), "whoops");
```

Comment:
```suggestion
    require(totalSupply == uint56(-1), "whoops");
```

Reply:
why? your suggestion would be caught by the test as `totalSupply` is currently always `uint(-1)`...

Reply:
I don't know why your require makes sense. There is no number larger than (-1) so your suggestion is never sat. And so it would revert for every input. My suggestion would make the test succeed for the setup we have, but no others

Reply:
I think you maybe negated this too many times? Or am I confused about what's going on?

Reply:
It's `uint56` not `uint256`. 

Looking at this again, it's probably not the best choice for an example 😄

I'll rework it to be more clear.

Reply:
Oh lol.. I see